File: typescript/packages/core/src/vfs/testing.ts
import type { Accessor } from '../accessor/base.ts'
import type { IndexCacheStore } from '../cache/index/store.ts'
import type { CommandIO } from '../commands/builtin/generic_bind/adapter.ts'
import { FileType, type PathSpec } from '../types.ts'
import { isEnoent } from '../utils/errors.ts'
import { VFSAdapter } from './adapter.ts'

/** A small known file, its parent directory, and an absent sibling. */
export interface ReadFixture {
  file: PathSpec
  directory: PathSpec
  missing: PathSpec
  content: Uint8Array
}

function check(condition: boolean, message: string): asserts condition {
  if (!condition) throw new Error(message)
}

function sameBytes(actual: Uint8Array, expected: Uint8Array): boolean {
  return actual.length === expected.length && actual.every((byte, i) => byte === expected[i])
}

/** Verify reads against a caller-owned fixture without mutations or test dependencies.
 * Native ranges probe nonempty windows within the fixture. Empty and out-of-range
 * reads are normalized by the filesystem operation instead. */
export async function checkReadContract<A extends Accessor>(
  adapter: VFSAdapter<A> | CommandIO<A>,
  accessor: A,
  fixture: ReadFixture,
  index?: IndexCacheStore,
): Promise<void> {
  const io = adapter instanceof VFSAdapter ? adapter.toCommandIO() : adapter
  const data = await io.readBytes(accessor, fixture.file, index)
  check(sameBytes(data, fixture.content), 'readBytes differs from fixture content')
  const info = await io.stat(accessor, fixture.file, index)
  check(info.type === FileType.FILE, 'stat must classify the fixture as a file')
  check(
    info.size === null || info.size === data.length,
    'stat size must be rendered byte length or null',
  )
  const parent = await io.stat(accessor, fixture.directory, index)
  check(parent.type === FileType.DIRECTORY, 'stat must classify the parent as a directory')
  const children = await io.readdir(accessor, fixture.directory, index)
  check(children.includes(fixture.file.virtual), 'readdir must include the child virtual path')
  const chunks: number[] = []
  for await (const chunk of io.readStream(accessor, fixture.file, index)) chunks.push(...chunk)
  check(sameBytes(Uint8Array.from(chunks), data), 'readStream differs from readBytes')
  if (io.readRange !== undefined && data.length > 0) {
    const offset = Math.min(1, data.length - 1)
    for (const size of [Math.min(3, data.length - offset), undefined]) {
      const actual = await io.readRange(accessor, fixture.file, index, offset, size ?? null)
      check(
        sameBytes(actual, data.slice(offset, size === undefined ? undefined : offset + size)),
        'readRange must use offset and byte count',
      )
    }
  }
  if (io.exists !== undefined) {
    check(await io.exists(accessor, fixture.file), 'exists rejected the fixture file')
    check(!(await io.exists(accessor, fixture.missing)), 'exists accepted a missing file')
  }
  for (const operation of [io.stat, io.readBytes]) {
    try {
      await operation(accessor, fixture.missing, index)
    } catch (error) {
      if (isEnoent(error)) continue
      throw error
    }
    throw new Error('missing paths must raise ENOENT')
  }
}

File: python/mirage/vfs/testing.py
from dataclasses import dataclass

from mirage.accessor.base import Accessor
from mirage.cache.index import NULL_INDEX, IndexCacheStore
from mirage.commands.builtin.generic_bind.adapter import CommandIO
from mirage.types import FileType, PathSpec
from mirage.vfs.adapter import VFSAdapter


@dataclass(frozen=True, slots=True)
class ReadFixture:
    """A small known file, its parent directory, and an absent sibling."""
    file: PathSpec
    directory: PathSpec
    missing: PathSpec
    content: bytes


async def check_read_contract(adapter: VFSAdapter | CommandIO,
                              accessor: Accessor,
                              fixture: ReadFixture,
                              index: IndexCacheStore = NULL_INDEX) -> None:
    """Verify reads against a caller-owned fixture without mutations.

    Args:
        adapter (VFSAdapter | CommandIO): adapter being validated.
        accessor (Accessor): fixture's backend client.
        fixture (ReadFixture): file, parent, absent sibling, and bytes.
        index (IndexCacheStore): cache to exercise, if any.
    """
    io = adapter.to_command_io() if isinstance(adapter,
                                               VFSAdapter) else adapter
    data = await io.read_bytes(accessor, fixture.file, index)
    assert data == fixture.content, "read_bytes differs from fixture content"
    info = await io.stat(accessor, fixture.file, index)
    assert info.type == FileType.FILE, "fixture must stat as a file"
    assert info.size is None or info.size == len(
        data), "stat size must be rendered byte length or None"
    parent = await io.stat(accessor, fixture.directory, index)
    assert parent.type == FileType.DIRECTORY, (
        "parent must stat as a directory")
    children = await io.readdir(accessor, fixture.directory, index)
    assert fixture.file.virtual in children, (
        "readdir must include the child virtual path")
    streamed = b"".join(
        [part async for part in io.read_stream(accessor, fixture.file, index)])
    assert streamed == data, "read_stream differs from read_bytes"
    if io.read_range is not None:
        for offset, size in [(0, 0), (1, 3), (len(data), 2), (1, None)]:
            end = None if size is None else offset + size
            actual = await io.read_range(accessor, fixture.file, index, offset,
                                         size)
            assert actual == data[
                offset:end], "read_range must use offset and byte count"
    if io.exists is not None:
        assert await io.exists(
            accessor, fixture.file), "exists rejected the fixture file"
        assert not await io.exists(
            accessor, fixture.missing), "exists accepted a missing file"
    for operation in (io.stat, io.read_bytes):
        try:
            await operation(accessor, fixture.missing, index)
        except FileNotFoundError:
            continue
        raise AssertionError("missing paths must raise FileNotFoundError")
